### PR TITLE
Make `ErrorSpec` check for all constructors, move golden tests together

### PR DESCRIPTION
**Pull request by carbolymer**
# Description
This is a follow-up PR after https://github.com/input-output-hk/cardano-api/pull/9#discussion_r1202675195

- [x] [Move golden tests together](https://input-output-rnd.slack.com/archives/G011N23CEAE/p1684859897519469?thread_ts=1684838032.919819&cid=G011N23CEAE)
- [x] [Add constructor names check in `ErrorSpec`](https://github.com/input-output-hk/cardano-api/pull/9#discussion_r1202675195)

# Changelog

```yaml
- description: |
    Make `ErrorSpec` check for all constructors. Move golden tests together.
  compatibility: no-api-changes
  type: test
```

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
  - integration tests
  See [Runnings tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [ ] The version bounds in `.cabal` files are updated
- [x] CI passes. See note on CI.  The following CI checks are required:
  - [x] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [x] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [x] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [x] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: cardano-api/test/cardano-api-golden/Test/Golden/Cardano/Api/Typed/Script.hs [hunk 1]
```diff
@@ -77,68 +86,60 @@ exampleSimpleScriptV2_MofN =
...
+  testPropertyNamed "roundtrip ScriptData" "roundtrip ScriptData" . H.property $ do
     sData <- H.forAll genHashableScriptData
     sData === fromAlonzoData (toAlonzoData @L.Alonzo sData)
 
-prop_roundtrip_HashableScriptData_JSON :: Property
-prop_roundtrip_HashableScriptData_JSON =
-  H.property $ do
+test_roundtrip_HashableScriptData_JSON :: TestTree
+test_roundtrip_HashableScriptData_JSON =
+  testPropertyNamed "roundtrip HashableScriptData"  "roundtrip HashableScriptData" . H.property $ do
```
Comment (Jimbo4350) [inline cardano-api/test/cardano-api-golden/Test/Golden/Cardano/Api/Typed/Script.hs:142]: We can eventually abstract this as a function: `testPropertyNamed "roundtrip HashableScriptData"  "roundtrip HashableScriptData" . H.property`

File: cardano-api/test/cardano-api-golden/Test/Golden/ErrorsSpec.hs [hunk 2]
```diff
@@ -345,9 +345,18 @@ test_TxMetadataRangeError =
 
 testAllErrorMessages :: forall a. (HasCallStack, Data a, Error a) => [a] -> TestTree
 testAllErrorMessages errs = withFrozenCallStack $ do
-  let typeName = show $ typeRep (Proxy @a)
+  let err = head errs
```
Comment (Jimbo4350) [inline cardano-api/test/cardano-api-golden/Test/Golden/ErrorsSpec.hs:348]: Although we don't expect to call `testAllErrorMessages` on an empty list, we should handle `head`'s potential exception explicitly as it's good practice.

**Review by Jimbo4350 — APPROVED** [review]
LGTM, just one comment regarding the use of `head`.

**Comment by newhoggy** [comment]
There's a build failure.

File: cardano-api/test/cardano-api-golden/Test/Golden/ErrorsSpec.hs (hunk 2 above)
Comment (carbolymer) [reply cardano-api/test/cardano-api-golden/Test/Golden/ErrorsSpec.hs:350]: `err`  is never evaluated here - we just need its `Data` instance here. The problem is with `dataTypeOf` - there is no equivalent function accepting `Proxy`.  I could use `undefined` here, but `head errs` seemed less of an eyesore.

**Review by newhoggy — APPROVED** [review]

**Review by Jimbo4350 — APPROVED** [review]

**Pull request merged**